src/Microsoft.IdentityModel.Protocol.Extensions/SamlSecurityTokenHandler.cs

```diff
@@ -287,118 +290,160 @@ public virtual ClaimsPrincipal ValidateToken(string securityToken, TokenValidati
+                    }
+                    else
                     {
-                        throw new SecurityTokenInvalidAudienceException(string.Format(CultureInfo.InvariantCulture, ErrorMessages.IDX10231, securityToken));
+                        issuer = ValidateIssuer(issuer, samlToken, validationParameters);
                     }
                 }
-                else
+
+                if (samlToken.Assertion.SigningToken != null)
                 {
-                    ValidateAudience(audiences, samlToken, validationParameters);
+                    ValidateIssuerSecurityKey(samlToken.Assertion.SigningToken.SecurityKeys[0], samlToken, validationParameters);
                 }
-            }
 
-            string issuer = null;
-            issuer = samlToken.Assertion.Issuer == null ? null : samlToken.Assertion.Issuer;
+                ClaimsIdentity identity = CreateClaimsIdentity(samlToken, issuer, validationParameters);
+                if (validationParameters.SaveSigninToken)
+                {
+                    identity.BootstrapContext = new BootstrapContext(securityToken);
+                }
 
-            if (validationParameters.ValidateIssuer)
+                validatedToken = samlToken;
+                return new ClaimsPrincipal(identity);
+            }
+            catch (SignatureVerificationFailedException ex)
             {
-                if (validationParameters.IssuerValidator != null)
+                XmlDocument xmlDoc = new XmlDocument();
+                xmlDoc.LoadXml(securityToken);
+                XmlNodeList keyInfoList = xmlDoc.GetElementsByTagName(KeyInfo);
+                if (keyInfoList.Count > 0
+                    && (validationParameters.IssuerSigningKey != null
+                    || validationParameters.IssuerSigningKeys != null
+                    || validationParameters.IssuerSigningToken != null
+                    || validationParameters.IssuerSigningTokens != null
+                    || validationParameters.IssuerSigningKeyResolver != null))
                 {
-                    issuer = validationParameters.IssuerValidator(issuer, samlToken, validationParameters);
+                    XmlNode keyInfoNode = keyInfoList.Item(0);
+                    if (String.IsNullOrEmpty(keyInfoNode.InnerXml))
+                    {
+                        // KeyInfo element is empty.
+                        throw;
+                    }
+                    else
+                    {
+                        if (issuerTokenResolver.IsKeyMatched)
+                        {
+                            // keyInfo in token matched with key(s) in validationParameters.
+                            throw;
+                        }
+                        else
+                        {
+                          throw new SecurityTokenSignatureKeyNotFoundException(string.Format(CultureInfo.InvariantCulture, ErrorMessages.IDX10506, securityToken), ex);
```

Comment:
indent a consistently. Add a comment that mostly likely means the user should refresh the key material.